File: src/main/java/Kotel/core/module/Module.java
package Kotel.core.module;

import Kotel.core.setting.Setting;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public abstract class Module {
    private final String name;
    private final String description;
    private final ModuleCategory category;
    private final List<Setting<?>> settings = new ArrayList<>();
    private boolean enabled;

    protected Module(String name, String description, ModuleCategory category) {
        this.name = name;
        this.description = description;
        this.category = category;
    }

    public final void toggle() { setEnabled(!enabled); }
    public final void setEnabled(boolean enabled) {
        if (this.enabled == enabled) return;
        this.enabled = enabled;
        if (enabled) onEnable(); else onDisable();
    }

    protected void onEnable() { }
    protected void onDisable() { }
    protected <T extends Setting<?>> T add(T setting) { settings.add(setting); return setting; }

    public String name() { return name; }
    public String description() { return description; }
    public ModuleCategory category() { return category; }
    public boolean enabled() { return enabled; }
    public List<Setting<?>> settings() { return Collections.unmodifiableList(settings); }
}

File: kotlin-mod/src/main/kotlin/com/koteldlc/client/modules/Module.kt
package com.koteldlc.client.modules

import com.google.gson.JsonElement
import com.google.gson.JsonObject

enum class Category {
    COMBAT,
    VISUAL
}

abstract class Module(
    val name: String,
    val category: Category,
    var enabled: Boolean = false
) {
    val settings: LinkedHashMap<String, Any> = linkedMapOf()

    open fun onEnable() {}
    open fun onDisable() {}
    open fun onTick() {}

    fun toggle() {
        enabled = !enabled
        if (enabled) onEnable() else onDisable()
    }

    fun applyEnabled(state: Boolean) {
        if (enabled != state) {
            toggle()
        }
    }

    fun setSetting(key: String, value: Any): Boolean {
        if (!settings.containsKey(key)) return false
        settings[key] = value
        return true
    }

    open fun save(): JsonObject {
        val json = JsonObject()
        json.addProperty("enabled", enabled)
        val settingsJson = JsonObject()
        settings.forEach { (k, v) ->
            when (v) {
                is Boolean -> settingsJson.addProperty(k, v)
                is Number -> settingsJson.addProperty(k, v)
                else -> settingsJson.addProperty(k, v.toString())
            }
        }
        json.add("settings", settingsJson)
        return json
    }

    open fun load(json: JsonObject) {
        if (json.has("enabled")) {
            applyEnabled(json.get("enabled").asBoolean)
        }
        val settingsJson = json.getAsJsonObject("settings") ?: return
        settings.forEach { (key, current) ->
            if (!settingsJson.has(key)) return@forEach
            val value: JsonElement = settingsJson.get(key)
            settings[key] = when (current) {
                is Boolean -> value.asBoolean
                is Int -> value.asInt
                is Double -> value.asDouble
                is Float -> value.asFloat
                is Long -> value.asLong
                else -> value.asString
            }
        }
    }
}
